hydra/experimental/initialize.py

```diff
@@ -46,6 +46,7 @@ def __init__(
             config_path=config_path,
             job_name=job_name,
             caller_stack_depth=caller_stack_depth + 1,
+            version_base=str(version.getbase()),
```

Comment:
Can you briefly explain why you had to add those `version_base=...` args in several places in this commit?

Reply:
Yes of course. First let me remark that I made those `version_base=...` changes to deprecated copies of the `initialize` classes that are stored in the `hydra.experimental.initialize` module. Most users will want to import from `hydra.initialize` instead of from `hydra.experimental.initialize`. If the client tries to use the classes from `hydra.experimental.instantiate`, they will get a `UserWarning` if they have set `version_base<=1.1` and they will get an `ImportError` if they set `version_base>=1.2` or `version_base is None`.

The `version_base=...` change above solves an issue related to testing those `UserWarnings`. Specifically, there are three tests that check this `UserWarning`/`ImportError` behavior, namely [`hydra.tests.test_compose.test_deprecated_{initialize,initialize_config_dir,initialize_config_module}`](https://github.com/facebookresearch/hydra/blob/d0b8e73ad52ba9fa8ff42ba1f28e2b03f216dff7/tests/test_compose.py#L662-L726). Those three tests were failing for reasons I'll describe below:

The deprecated classes in `hydra.experimental.initialize` work by first issuing a warning or error (depending on the `version_base` setting) and then delegating to the corresponding class in the `hydra.initialize` module. The problem is that the `__init__` methods on the classes in `hydra.experimental.initialize` have slightly different signatures than the corresponding methods in `hydra.initialize`. In particular, the classes in `hydra.initialize` take a `version_base` keyword argument while the classes in `hydra.experimental.initialize` do not. Before the changes made in this PR, the classes in `hydra.experimental.initialize` would not pass a `version_base` keyword argument to the delegate, which resulted in the default keyword argument `version_base=hydra.initialize._UNSPECIFIED_` being used by the `__init__` methods on the classes in `hydra.initialize` module. This in turn caused an additional `UserWarning` to be raised (complaining that "`The version_base parameter is not specified`") in addition to the `UserWarning` under test (which was supposed to complain that the classes in `hydra.experimental.initialize` are deprecated). The extra `UserWarning`s were causing those three tests to fail.

As far as why the presence of two `UserWarnings` (instead of just the one `UserWarning` that was under test) caused CI to fail recently but did not cause failures originally: I believe this is related to differences between pytest v7.x and pytest v8.x in terms of how warnings are handled. I haven't investigated fully to confirm that this is the case, however.

Reply:
Got it, thanks! Sorry it ended up being a pretty long explanation ;)